### ISSUE TITLE
Bug: VSCode Latest Release Breaks MD Preview

### ISSUE THREAD

**Issue by joshuatz** [joshuatz/vscode-do-co-authoring#2]
The latest release ([June 2021 / `v 1.58`](https://code.visualstudio.com/updates/v1_58)) is breaking part of the Markdown-it plugin / preview transformation, due to the new [KaTeX support](https://code.visualstudio.com/updates/v1_58#_math-formula-rendering-in-the-markdown-preview). Specifically, the notes are failing to render properly, likely due to a conflict with the `<$>` delimiters and the KaTeX token recognizer (`$_____$` is used as inline math syntax).

![2021-07-15 14_31_28-Window](https://user-images.githubusercontent.com/17817563/125860518-93687aab-62f6-440e-99a9-adf1a6374dd1.png)

**Comment by joshuatz** [comment]
Some additional notes:

- Tracked in VSCode as [#124783](https://github.com/microsoft/vscode/issues/124783) (feature request), and [here is commit that brought it in](https://github.com/microsoft/vscode/commit/0a57fd87b1d1ef0ff81750f84840ee4303b8800b)
- Internally, VSCode just [loads and uses an external Markdown-it plugin](https://github.com/microsoft/vscode/blob/2494e3a1f08bcb925a563b661b901ff9c26a3507/extensions/markdown-math/src/extension.ts#L22-L27), [@iktakahiro/markdown-it-katex](https://github.com/iktakahiro/markdown-it-katex). This adds named rules of:
    - `math_inline`
    - `math_block`

I'm a bit surprised that this is happening, since I already am disabling `math_inline` as part of my loading strategy. I wonder if the order of rule loading changed...

**Comment by joshuatz** [comment]
Ah, `math_inline` is now being loaded twice because I have [the `Markdown All in One` extension](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one), which also [adds math support](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one#math) with KaTeX, with a named rule of `math_inline`.

When I call `md.disable('math_inline', true)`, it is only disabling the first of the two `math_inline` instances. This kinda seems like a a bug with `Markdown-it`, because it seems like once > 2 rules exist with the same name in a given *ruler*, there is no clear way to disable more than just the first one, other than by drilling through MDIT internals. I won't step through every part of the Markdown-it internals, but to give a quick explanation, the problem is that the `.disable()` methods work [by getting the index of the rule with the passed in name](https://github.com/markdown-it/markdown-it/blob/e5986bb7cca20ac95dc81e4741c08949bf01bb77/lib/ruler.js#L319-L325), by using [the Ruler `__find__` method](https://github.com/markdown-it/markdown-it/blob/e5986bb7cca20ac95dc81e4741c08949bf01bb77/lib/ruler.js#L48-L57), which is problematic, because that assumes there is only one rule matching a given name.

I can confirm this easily with the following code, which bypasses the public APIs for MDIT, and disables all matching rules directly:

```ts
md.inline.ruler.__rules__.forEach(r => {
	if (r.name === 'math_inline') {
		r.enabled = false;
	}
})
md.inline.ruler.__cache__ = null;
```

## Going Forward
Couple of ways I could address this:

- Encourage users not to use conflicting extensions (not a fan of this approach)
- Start using a custom WebView (also related to #1)
- Use code similar to above, to disable all possible instances of `math_inline` rule
- See if I can change the loading logic for my *own* rules, to make sure they load first and transform the `<$>` tokens before they can be parsed by the math rules.
    - Edit: This is harder then it sounds, since there is no public `unshift()` method for adding rules with MDIT.

**Comment by joshuatz** [comment]
I ended up going with option C, disabling all rule instances of `math_inline`. I justify this by the fact that users explicitly opt-in to having extension applied to previews, and these rules are not supported by DOCO anyways, so letting them run could possibly mislead users.

This is now fixed / closed, but is also another tick-mark in the column of "why I should consider just building a custom Webview pane for this".
